server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -476,21 +480,21 @@ private static boolean containsSnippet(Map<?, ?> map, String snippet) {
 
     private static boolean containsSnippet(List<?> list, String snippet) {
         for (Object value : list) {
-            if (value instanceof Map) {
-                if (containsSnippet((Map<?, ?>) value, snippet)) {
-                    return true;
-                }
-            } else if (value instanceof List) {
-                if (containsSnippet((List<?>) value, snippet)) {
-                    return true;
-                }
-            } else if (value instanceof String) {
-                String valueString = (String) value;
-                if (valueString.contains(snippet)) {
-                    return true;
-                }
+            if (containsSnippet(value, snippet)) {
+                return true;
             }
         }
         return false;
     }
+
+    private static boolean containsSnippet(Object value, String snippet) {
+        if (value instanceof Map) {
+            return containsSnippet((Map<?, ?>) value, snippet);
+        } else if (value instanceof List) {
+            return containsSnippet((List<?>) value, snippet);
+        } else if (value instanceof String) {
+            return ((String) value).contains(snippet);
+        }
+        return false;
+    }
```

Comment:
This all makes me think that we need to extend the XContentMapValues functions to take predicates or lambdas or something, we have a lot of repeated 'descend through json objects' code here and elsewhere.  But that can be for later, we don't need it for this PR.

Reply:
true I have refactored two places where we do that as part of this PR.